### PR TITLE
Feat: remove envbinding controller

### PR DESCRIPTION
**Pull request by Somefive**
#### Major Change
- make envbinding completely implemented in workflowstep, including 3 major steps:
  a. placement (decide which cluster or namespace to deploy)
  b. patch (apply patch and component selector to generate app for env)
  c. apply (dispatch resources)

The placement decision (which cluster and namespace to use for the env) is stored in application status. Other variables such as PatchedApp or ApplicationComponent are purely in memory use (used between workflowSteps).

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [x] Run `make reviewable` to ensure this PR is ready for review.

### Special notes for your reviewer

Since the EnvBinding Policy spec is not changed, only the underlying mechanism is reworked. It is completely backward-compatible.

Several issues will be addressed after this PR:
1. Support Add/Edit/Delete envs. (Major problem here is GC.)
2. Solve revisions in sub-cluster do not accept override namespace. (Major problem here is also GC.)
3. Support cluster label selector & namespace label selector. (Since the placement decision now is embedded in AppStatus, it is possible to achieve that now.)

### CODE REVIEW

**Review by leejanee — APPROVED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 1]
```diff
@@ -353,7 +356,14 @@ func (r *Reconciler) endWithNegativeCondition(ctx context.Context, app *v1beta1.
...
 	updateObservedGeneration(app)
-	return r.Client.Status().Patch(ctx, app, client.Merge)
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		status := app.Status.DeepCopy()
+		if err := r.Client.Get(ctx, client.ObjectKeyFromObject(app), app); err != nil {
+			return err
+		}
+		app.Status = *status
+		return r.Status().Patch(ctx, app, client.Merge)
+	})
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:366]: This should be a separated PR

Comment (Somefive) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:366]: ok

**Review by yangsoon — APPROVED** [review]

File: pkg/stdlib/pkgs/multicluster.cue [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	envConfig: envMap["\(env)"]
+
+	placementDecisions: multicluster.#MakePlacementDecisions & {
+		inputs: {
+			policyName: policy
+			envName:    env
+			placement:  envConfig.placement
+		}
+	} @step(3)
```
Comment (FogDong) [inline pkg/stdlib/pkgs/multicluster.cue:94]: missing @step(2)

**Review by FogDong — APPROVED** [review]

File: pkg/stdlib/pkgs/multicluster.cue (hunk 2 above)
Comment (Somefive) [reply pkg/stdlib/pkgs/multicluster.cue]: ok

**Review by zzxwill — APPROVED** [review]

**Pull request merged**